common/kafka/kafka_consumer.cpp

```diff
@@ -20,293 +20,248 @@
+      consumer_->resetInstance(nullptr);
+    }
+
+    // TODO: Change to unique_ptr
+    std::vector<std::shared_ptr<RdKafka::TopicPartition>> topic_partitions;
+    // A temporary RdKafka::TopicPartition* vector specifically used
+    // for KafkaConsumer->assign(). The pointers inside this vector is
+    // owned by `topic_partitions`.
+    std::vector<RdKafka::TopicPartition*> tmp_topic_partitions;
+    topic_partitions.reserve(partition_ids_.size() * topic_names_.size());
+    tmp_topic_partitions.reserve(partition_ids_.size() * topic_names_.size());
+    for (const auto partition_id : partition_ids_) {
+      for (const auto& topic_name : topic_names_) {
+        topic_partitions.push_back(
           std::shared_ptr<RdKafka::TopicPartition>(
-              RdKafka::TopicPartition::create(
+            RdKafka::TopicPartition::create(
               topic_name,
               partition_id,
               // Use the end of the offset to init here, but it does not matter.
               // When kafkaconsumer is used, the caller is supposed to call
               // seek() to the expected offset before consuming messages.
               RdKafka::Topic::OFFSET_END)));
-      tmp_topic_partitions.push_back(topic_partitions.back().get());
-      topic_names_string_.append(topic_name);
-      topic_names_string_.append(", ");
+        tmp_topic_partitions.push_back(topic_partitions.back().get());
+        topic_names_string_.append(topic_name);
+        topic_names_string_.append(", ");
+      }
     }
-  }
-  const auto error_code = ExecuteKafkaOperationWithRetry(
+    const auto error_code = ExecuteKafkaOperationWithRetry(
       [this, &tmp_topic_partitions]() {
-        return consumer_->assign(tmp_topic_partitions); });
-  if (error_code != RdKafka::ERR_NO_ERROR) {
-    LOG(ERROR) << "Failed to assign partitions: " << partition_ids_str_
-               << ", error_code: " << RdKafka::err2str(error_code)
-               << ", num retries: " << FLAGS_kafka_consumer_num_retries;
-    common::Stats::get()->Incr(getFullStatsName(
+        return consumer_->getInstance()->assign(tmp_topic_partitions);
+      });
+
+    if (error_code != RdKafka::ERR_NO_ERROR) {
+      LOG(ERROR) << "Failed to assign partitions: " << partition_ids_str_
+                 << ", error_code: " << RdKafka::err2str(error_code)
+                 << ", num retries: " << FLAGS_kafka_consumer_num_retries;
+      common::Stats::get()->Incr(getFullStatsName(
         kKafkaConsumerErrorAssign,
         {kafka_consumer_type_metric_tag_,
          "error_code=" + std::to_string(error_code)}));
-    return;
+
+      if (std::atomic_load(&reset_)) {
+        continue;
+      }
+
+      return;
+    }
+    LOG(INFO) << "Successfully created kafka consumer for partitions: "
+              << partition_ids_str_ << ", topics: "
+              << folly::join(",", topic_names_);
+    is_healthy_.store(true);
+  } while (std::atomic_load(&reset_));
```

Comment:
Is this while loop for the case where the callback is called while the partitions are being assigned?

Reply:
while loop is to ensure the operations within single call of a method happens without any updates on file changes as notified...